java/arcs/core/storage/StorageProxy.kt

```diff
@@ -63,6 +72,33 @@ class StorageProxy<Data : CrdtData, Op : CrdtOperationAtTime, T>(
     private val store: StorageCommunicationEndpoint<Data, Op, T> =
         storeEndpointProvider.getStorageEndpoint(ProxyCallback(::onMessage))
 
+    private val outgoingMessagesChannel =
+        Channel<Pair<ProxyMessage<Data, Op, T>, CompletableDeferred<Boolean>>>(Channel.UNLIMITED)
+
+    init {
+        // Send messages generated by applyOp to the Store in the order in which they were received
+        // by:
+        //
+        // 1. Calling store.onProxyMessage() with the message for each message.
+        // 2. Buffering those calls so they happen in parallel/pipelined-with calling store.idle()
+        // 3. Once the store is self-identifying as idle, complete the deferred which was generated
+        //    at the same time as the message (and was returned to the caller)
+        outgoingMessagesChannel.consumeAsFlow()
+            .map { (message, deferredToComplete) ->
+                log.debug { "Sending operations to store" }
+                store.onProxyMessage(message)
+                log.debug { "Operations sent to store" }
+                deferredToComplete
+            }
+            .buffer()
+            .onEach {
+                withTimeoutOrNull(1000) { store.idle() }
+                it.complete(true)
+            }
+            .flowOn(Dispatchers.Default)
+            .launchIn(scheduler.scope)
```

Comment:
when StorageProxy.close() is invoked, should we cancel/close this channel? I'm wondering if the scheduler is cancelled, but the channel isn't closed, if the internal buffer will end up retaining the models?

Reply:
it might retain the models, but the scheduler and proxy shoudl be gone - so there shouldn't be a route to the GC root, right?